src/jit/lclvars.cpp

```diff
@@ -1563,9 +1565,16 @@ void Compiler::lvaCanPromoteStructType(CORINFO_CLASS_HANDLE    typeHnd,
                     return;
                 }
 
-                // Non-primitive struct field.
-                // Try to promote structs of single field of scalar types aligned at their
-                // natural boundary.
+                // Try to promote byref-like structs with a single field of scalar type that is aligned at its
+                // type's natural boundary.
+
+                // Do not promote if the single field is not byref-like.
+                const DWORD fieldFlags       = info.compCompHnd->getClassAttribs(pFieldInfo->fldTypeHnd);
+                const bool  fieldIsByRefLike = (fieldFlags & CORINFO_FLG_CONTAINS_STACK_PTR) != 0;
+                if (!fieldIsByRefLike)
+                {
+                    return;
+                }
 
```

Comment:
Happily, that case seems to fall out due to "normed structs": for structs with a single pointer-sized integral field, the VM reports the type of the struct as the type of the field and we are able to perform the promotion. Here is an example:

`struct T` is normalized to `int`:
```
struct S
{
    struct T
    {
        public int i;
    }

    T t;
    int j;

    static int Main(string[] args)
    {
        S s;
        s.t.i = 42;

        if (args.Length == 0)
        {
            S s2;
            s2.t.i = 100;
            s2.j = 0;

            s = s2;
        }

        return s.t.i;
    }
}
```
```
*************** In fgPromoteStructs()

lvaTable before fgPromoteStructs
; Initial local variable assignments
;
;  V00 arg0              ref  class-hnd
;  V01 loc0           struct ( 8)  ld-addr-op
;  V02 loc1           struct ( 8)  ld-addr-op

Promoting struct local V01 (S):
lvaGrabTemp returning 3 (V03 tmp0) (a long lifetime temp) called for field V01.t (fldOffset=0x0).

lvaGrabTemp returning 4 (V04 tmp1) (a long lifetime temp) called for field V01.j (fldOffset=0x4).

Promoting struct local V02 (S):
lvaGrabTemp returning 5 (V05 tmp2) (a long lifetime temp) called for field V02.t (fldOffset=0x0).

lvaGrabTemp returning 6 (V06 tmp3) (a long lifetime temp) called for field V02.j (fldOffset=0x4).

lvaTable after fgPromoteStructs
; Initial local variable assignments
;
;  V00 arg0              ref  class-hnd
;  V01 loc0           struct ( 8)  ld-addr-op
;  V02 loc1           struct ( 8)  ld-addr-op
;  V03 tmp0              int  V01.t(offs=0x00) P-INDEP
;  V04 tmp1              int  V01.j(offs=0x04) P-INDEP
;  V05 tmp2              int  V02.t(offs=0x00) P-INDEP
;  V06 tmp3              int  V02.j(offs=0x04) P-INDEP
```

`struct T` is not normalized:
```
struct S
{
    struct T
    {
        public object i;
    }

    T t;
    int j;

    static int Main(string[] args)
    {
        S s;
        s.t.i = 42;

        if (args.Length == 0)
        {
            S s2;
            s2.t.i = 100;
            s2.j = 0;

            s = s2;
        }

        return (int)s.t.i;
    }
}
```

With this change:
```
lvaTable before fgPromoteStructs
; Initial local variable assignments
;
;  V00 arg0              ref  class-hnd
;  V01 loc0           struct ( 8)  ld-addr-op
;  V02 loc1           struct ( 8)  ld-addr-op
;  V03 tmp0              ref 
;  V04 tmp1              ref 

lvaTable after fgPromoteStructs
; Initial local variable assignments
;
;  V00 arg0              ref  class-hnd
;  V01 loc0           struct ( 8)  ld-addr-op
;  V02 loc1           struct ( 8)  ld-addr-op
;  V03 tmp0              ref 
;  V04 tmp1              ref 
```

Without this change:
```
lvaTable before fgPromoteStructs
; Initial local variable assignments
;
;  V00 arg0              ref  class-hnd
;  V01 loc0           struct ( 8)  ld-addr-op
;  V02 loc1           struct ( 8)  ld-addr-op
;  V03 tmp0              ref 
;  V04 tmp1              ref 

Promoting struct local V01 (S):
lvaGrabTemp returning 5 (V05 tmp2) (a long lifetime temp) called for field V01.j (fldOffset=0x0).

lvaGrabTemp returning 6 (V06 tmp3) (a long lifetime temp) called for field V01.t (fldOffset=0x4).

Promoting struct local V02 (S):
lvaGrabTemp returning 7 (V07 tmp4) (a long lifetime temp) called for field V02.j (fldOffset=0x0).

lvaGrabTemp returning 8 (V08 tmp5) (a long lifetime temp) called for field V02.t (fldOffset=0x4).

lvaTable after fgPromoteStructs
; Initial local variable assignments
;
;  V00 arg0              ref  class-hnd
;  V01 loc0           struct ( 8)  ld-addr-op
;  V02 loc1           struct ( 8)  ld-addr-op
;  V03 tmp0              ref 
;  V04 tmp1              ref 
;  V05 tmp2              int  V01.j(offs=0x00) P-INDEP
;  V06 tmp3              ref  V01.t(offs=0x04) P-INDEP
;  V07 tmp4              int  V02.j(offs=0x00) P-INDEP
;  V08 tmp5              ref  V02.t(offs=0x04) P-INDEP
```